test/toolchain/toolchain_test.bzl

```diff
@@ -0,0 +1,109 @@
+"""End to end tests for rust toolchains."""
+
+load("@bazel_skylib//lib:unittest.bzl", "analysistest", "asserts")
+load("@bazel_skylib//rules:write_file.bzl", "write_file")
+load("//rust:defs.bzl", "rust_library")
+load("//rust:toolchain.bzl", "rust_stdlib_filegroup", "rust_toolchain")
+
+EXEC_TOOLCHAIN_FLAG = "missing"
+TOOLCHAIN_FLAG = "before"
+CONFIG_FLAG = "after"
+
+def _toolchain_adds_rustc_flags_impl(ctx):
+    """Checks that a library built with a toolchain that specifys extra_rustc_flags will
+    have those flags appended to its argv. Also makes sure that if the extra_rustc_flags
+    config is used, then those flags will end up AFTER the toolchain flags"""
+    env = analysistest.begin(ctx)
+
+    target = analysistest.target_under_test(env)
+    action = target.actions[0]
+    asserts.equals(env, "Rustc", action.mnemonic)
+
+    asserts.true(
+        env,
+        action.argv[-2:] == [TOOLCHAIN_FLAG, CONFIG_FLAG],
+        "Unexpected rustc flags: {}\nShould have ended with: {}".format(
+            action.argv,
+            [TOOLCHAIN_FLAG, CONFIG_FLAG],
+        ),
+    )
+
+    asserts.true(
+        env,
+        EXEC_TOOLCHAIN_FLAG not in action.argv,
+        "Found exec toolchain flag ({}) in rustc flags: {}".format(EXEC_TOOLCHAIN_FLAG, action.argv),
+    )
+
+    return analysistest.end(env)
+
+toolchain_adds_rustc_flags_test = analysistest.make(
+    _toolchain_adds_rustc_flags_impl,
+    config_settings = {
+        "//command_line_option:extra_toolchains": ["@rules_rust//test/toolchain:extra_flags_toolchain"],
```

Comment:
Did a bit of debugging with @scentini and suspect the issue is that here we're overriding the custom `--extra_toolchains=@buildkite_config//config:cc-toolchain` part that's passed in RBE builds (https://buildkite.com/bazel/rules-rust-rustlang/builds/7469#01848db5-aa45-46dc-b12d-027efe604c8e/237-240).
We could try and either always add this here (not sure what will happen in non-RBE environments) or bring back the user-defined transition where we can do things like: append the custom rust toolchain to the parent `//command_line_option:extra_toolchains` list

Reply:
setting the buildkite toolchain in the config_settings doesn't seem to work locally. I still don't understand where exactly the transition should go.

If I make a wrapper rule that sets `cfg = my_transition` and a target using that rule depends on the `rust_library`, then building the wrapper target will trigger the transition and build the library with the extra configs from the transition. But then the `target_under_test` will be that wrapper target instead of the library, and i can't figure out how to get the actions of the library target in that case.

Reply:
PTAL at https://github.com/bazelbuild/rules_rust/blob/main/test/unit/is_proc_macro_dep/is_proc_macro_dep_test.bzl

We create a [`custom rule`](https://github.com/bazelbuild/rules_rust/blob/main/test/unit/is_proc_macro_dep/is_proc_macro_dep_test.bzl#L42) for testing, that applies the [`transition`](https://github.com/bazelbuild/rules_rust/blob/main/test/unit/is_proc_macro_dep/is_proc_macro_dep_test.bzl#L36), and that attaches an [`aspect`](https://github.com/bazelbuild/rules_rust/blob/main/test/unit/is_proc_macro_dep/is_proc_macro_dep_test.bzl#L18) to the `dep` attribute that collects the actions for the dependency (which will be a `rust_library`). The `target_under_test` will be the custom rule we created, and we will inspect the actions collected by the aspect: https://github.com/bazelbuild/rules_rust/blob/main/test/unit/is_proc_macro_dep/is_proc_macro_dep_test.bzl#L128

I know this is all very complicated, but it's the best that the unit testing framework provides at this time.

Reply:
Thanks so much! I never would have thought to use an aspect to collect the action through the transition rule wrapper. Looks like it passes in CI as well now.